### ISSUE TITLE
buildConfigProjectForJob does not loop through all namespaces

### ISSUE THREAD

**Issue by coreydaley** [openshift/jenkins-sync-plugin#200]
The buildConfigProjectForJob method in the PipelineJobListener class does not correctly loop through all namespaces as both paths of the if/else have a return statement.

https://github.com/openshift/jenkins-sync-plugin/blob/master/src/main/java/io/fabric8/jenkins/openshiftsync/PipelineJobListener.java#L199-L213

**Comment by gabemontero** [comment]
@coreydaley @bparees - as part of this do we also want to deal with the `replace` vs. `patch` notion on the BC change per the "delete runs for corresponding builds" PR discussion ?

**Comment by bparees** [comment]
no, patch vs replace is a separate issue (unless i'm missing something).

**Comment by gabemontero** [comment]
ok ... then please open a tracking issue for that as well @coreydaley

**Comment by coreydaley** [comment]
> ok ... then please open a tracking issue for that as well @coreydaley


Opened https://github.com/openshift/jenkins-sync-plugin/issues/201

**Comment by gabemontero** [comment]
@coreydaley assigned to you - let me know if you're plate is too full with other items, etc.

**Comment by andreaseger** [comment]
Hi is there any update to this? We'd like to load imagestreams (for jenkins slaves) from multiple projects and this bug is preventing us from doing so.

**Comment by bparees** [comment]
> Hi is there any update to this? We'd like to load imagestreams (for jenkins slaves) from multiple projects and this bug is preventing us from doing so.

@andreaseger can you explain what issue this is causing for you?

@coreydaley my understanding of your PR for "fixing" this just changes the logic, it doesn't change the behavior (which is to create the buildconfig in the first namespace configured in the sync plugin).

**Comment by coreydaley** [comment]
@bparees that is correct, it just makes the code more clear about what it is doing without having to wonder why the loop only executes once.

**Comment by andreaseger** [comment]
we'd like to have our custom jenkins slave images in a single namespace on OS but use them in many jenkins instance in many namespaces.

So we have a namespace `custom-jenkins-salves` in OS which basically just has imagestreams for custom jenkins-slaves (correctly annotated/labeled so this plugin would find them)

And we have many other namespaces which spawn their own jenkins instance and run tests/trigger builds for various things. These namespaces do not by themselves have image-streams for our custom jenkins slaves. 

In theory we could simply add both projects/namespaces in the settings for this plugin. But due to this bug only the first namespace is actually synced. So it either finds/syncs the pipelines from the "owning" namespace OR syncs the global custom jenkins-slave imagestreams.

At least to me it looks like the bug mentioned here is related to the issue we're seeing.

**Comment by coreydaley** [comment]
@andreaseger What you described does not sound like what this bug is for.  
From the pull request:
> The PipelineJobListener takes jobs created in Jenkins that match a specific jobNamePattern
> and creates buildConfigs for them in OpenShift. It also updates the buildConfig in
> OpenShift if the job is updated in Jenkins
It sounds like you are taking about issues of syncing from OpenShift -> Jenkins, not the other way around.  Is that correct?

**Comment by bparees** [comment]
> At least to me it looks like the bug mentioned here is related to the issue we're seeing.

it's a bit confusing but what is described in this issue is not a bug.  it's just unclear code.

> So we have a namespace custom-jenkins-salves in OS which basically just has imagestreams for custom jenkins-slaves (correctly annotated/labeled so this plugin would find them)

You want the sync plugin to watch for imagestreams/configmaps in multiple potential namespaces, labeled as slaves.  I would expect that should already be happening (when you configure which namespaces the sync plugin watches for buildconfigs to run, that should also be configuring it for which namespaces to watch for configmaps/imagestreams to pick up slave definitions).  @gabemontero can confirm/deny that though.  Maybe we only look in the namespace jenkins is running in, for the configmaps/imagestreams.

Regardless, it's not this issue.

**Comment by gabemontero** [comment]
I can confirm the configmap / imagestream watches iterate over the namespace array from the sync plugin config.  Watches for each of those api resources are created across each of the configured namespaces.

**Comment by andreaseger** [comment]
hmm ok then I missunderstood the issue/description. Somehow I had issues with the syncing of multiple namespaces.

What I tried is having a namspace `jenkins` which has the slave image streams and a `playground` namespace where the jenkins and the buildconfig/pipeline is defined.

If I set in the Jenkins config "OpenShift Jenkins Sync" |> "Namespace" to `jenkins playground`.

It finds the imagestreams from `jenkins` and adds the kubernetes pod templates. It also created a Pipeline project for the bc in `playground`. But if I use "start pipeline" in OS it seems like nothing is happening in jenkins - turns out I was not patient enough and after a few minutes the build would be started in jenkins. I can see a 5 minutes delay here.

If I change the order in the settings to `playground jenkins` then clicking on "start pipeline" in openshift almost immediately triggers the build in jenkins.

Although this is unrelated to the issue at hand now(or is it?) and the syncing indeed seems to work as expected can you explain why `playground jenkins` is nearly instant while `jenkins playground` has a huge delay?

**Comment by bparees** [comment]
@andreaseger can you open a new issue for what you're observing?  I know @gabemontero has hunted down some similar issues w/ delays in jobs starting.

**Comment by andreaseger** [comment]
ok will do, thanks for the help so far
